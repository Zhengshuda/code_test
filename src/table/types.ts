/* eslint-disable @typescript-eslint/no-explicit-any */

import { InjectionKey, Ref } from '@vue/composition-api'
import { VNode } from 'vue'
import type { PropOptions, PropType } from 'vue-types/dist/types'
type Prop<T, D = T> = PropOptions<T, D> | PropType<T>
type PublicRequiredKeys<T> = {
  [K in keyof T]: T[K] extends { required: true } ? K : never
}[keyof T]

type PublicOptionalKeys<T> = Exclude<keyof T, PublicRequiredKeys<T>>
type InferPropType<T> = T extends null
  ? any // null & true would fail to infer
  : T extends { type: null | true }
    ? any // As TS issue https://github.com/Microsoft/TypeScript/issues/14829 // somehow `ObjectConstructor` when inferred from { (): T } becomes `any` // `BooleanConstructor` when inferred from PropConstructor(with PropMethod) becomes `Boolean`
    : T extends ObjectConstructor | { type: ObjectConstructor }
      ? Record<string, any>
      : T extends BooleanConstructor | { type: BooleanConstructor }
        ? boolean
        : T extends Prop<infer V, infer D>
          ? unknown extends V
            ? D
            : V
          : T

// eslint-disable-next-line @typescript-eslint/ban-types
export type IxPublicPropTypes<O> = O extends object
  ? { [K in PublicRequiredKeys<O>]: InferPropType<O[K]> } & { [K in PublicOptionalKeys<O>]?: InferPropType<O[K]> }
  : { [K in string]: any }

// 自定义props声明
export type Align = 'left' | 'right' | 'center';
export type SortDirection = '' | 'asc' | 'desc' | 'ASC' | 'DESC' | undefined;
export type TableColumn = {
  title?: string,
  key: string,
  align?: Align,
  sort?: boolean,
  sortDirection: SortDirection,
  className?: string,
  render?: (data: DataType, index: number) => VNode,
  width?: number | string,
  sortFn?: SortFnType<any>
};
export type TableColumns = Array<TableColumn>;
export type DataType = Record<string, any>;
export type SortFnDirection = 'ASC' | 'DESC';
export type SortFnType<T extends any> = (direction: SortFnDirection, a: T, b: T) => number;
export type PageNationType = {
  /* 页面总数 */
  totalPage: number,

  /* 数据总数 */
  totalData: number,

  /* 当前页数 */
  page: number,

  /* 每页显示的数量 */
  size: number,
};

export interface TableKeyInjection {
  columnsRef: Ref<TableColumns>
  headerAlign: Ref<Align>,
  align: Ref<Align>,
  dataRef: Ref<DataType[]>,
  originData: DataType[],
  rowHeight: Ref<string | number>,
  originPageRef: Ref<PageNationType>,
  sortFunction: (e: MouseEvent, column: TableColumn, sortDirection: Ref<SortDirection>) => void,
  lastPagenationStep: () => void,
  nextPagenationStep: () => void,
  setPagenationStep: (pageData: Partial<PageNationType>) => void
}

export interface TableBodyRef {
  sortData: Function
}

// Props 定义在这里
export const tableProps = {
  columns: {
    type: Array as PropType<TableColumns>,
    default: () => [],
    required: true
  },
  data: {
    type: Array as PropType<DataType[]>,
    default: () => []
  },
  headerAlign: {
    type: String as PropType<Align>,
    default: '',
  },
  align: {
    type: String as PropType<Align>,
    default: '',
  },
  border: {
    type: Boolean,
    default: true
  },
  rowHeight: {
    type: [Number, String],
    default: 40
  },
  pagenation: {
    type: Object as PropType<PageNationType>
  }
}

export type TablePublicProps = IxPublicPropTypes<typeof tableProps>

export const TableKey: InjectionKey<TableKeyInjection> = Symbol('Table')