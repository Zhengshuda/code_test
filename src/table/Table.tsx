/**
 * Created by uedc on 2021/10/11.
 */

import {
  computed,
  defineComponent,
  provide,
} from '@vue/composition-api'
import { TableKey, TableKeyInjection, tableProps } from './types'
import TableContent from './components/Content'
import TablePagenation from './components/Pagenation'
import { useData } from './hook/use-data'
import { useColumns } from './hook/use-columns'
import { usePage } from './hook/use-page'
import { Logger } from '@/utils/Logger'
import { isFunction } from 'lodash'
import { useSorter } from './hook/use-sorter'
import { useTableSorter } from './hook/use-table-sorter'

export default defineComponent({
  name: 'Table',
  props: tableProps,
  emits: ['page-change'],
  setup(props, { emit }) {
    Logger.trace('Table.tsx/setup', '传入表格组件的参数', props)

    if (!Array.isArray(props.data)) {
      Logger.warn('useData', '传入的data不能为非数组', props.data)
    }

    // 初始数据
    const originDataRef = computed(() => Array.isArray(props.data) ? props.data : [])

    // 分页数据
    const pagenationRef = computed(() => props.pagenation)

    // 调用useColumns，获取一个新的数组
    const columnsRef = useColumns(props)

    // 获取初始化排序状态
    const initedState = useTableSorter(columnsRef)

    // 调用useSorter，获取排序状态
    const {
      sortStateRef,
      changeSortState,
    } = useSorter(initedState)

    // 调用usePage，获取分页状态
    const {
      originPageRef,
      lastPagenationStep,
      nextPagenationStep,
      setPagenationStep,
    } = usePage(originDataRef, pagenationRef, emit)

    // 调用useData，获取表格数据
    const {
      dataRef,
    } = useData(
      columnsRef,
      pagenationRef,
      originPageRef,
      sortStateRef,
      originDataRef)

    const privideData: TableKeyInjection = {
      columnsRef: columnsRef,
      align: computed(() => props.align),
      contentBorder: computed(() => props.contentBorder),
      dataRef,
      originPageRef,
      sortState: computed(() => sortStateRef.value),
      changeSortState,
      lastPagenationStep,
      nextPagenationStep,
      setPagenationStep,
      emitFn: emit,
    }

    Logger.debug('Table.tsx/setup', '通过hook后进入组件的变量', privideData)
    provide(TableKey, privideData)

    return () => (
      <div class={{
        'table-wrap': true,
        'table-wrap-border': props.border,
        'table-wrap-border-has-pagenation': props.pagenation,
      }}>
        <TableContent></TableContent>
        {
          props.data.length ? (
            props.pagenation ? (
              <div class="table-pagenation">
                <TablePagenation
                  pageConfig={originPageRef.value}
                  lastPagenationStep={lastPagenationStep}
                  nextPagenationStep={nextPagenationStep}
                  setPagenationStep={setPagenationStep}></TablePagenation>
              </div>
            ) : null
          ) : (
            isFunction(props.emptyRender) ?
              props.emptyRender() :
              props.emptyRender
          )
        }
      </div>
    )
  },
})
