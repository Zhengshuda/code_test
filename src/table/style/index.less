.table-wrap {
  width: 100%;

  &.table-wrap-border {
    border: 1px solid rgba(239, 239, 245, 1);
    border-bottom: 0;

    &.table-wrap-border-has-pagenation {
      border-bottom: 1px solid rgba(239, 239, 245, 1);
    }
  }

  .table-content {
    width: 100%;
    box-sizing: border-box;
    border-spacing: 0;

    .table-thead {
      width: 100%;
      background-color: rgb(249, 249, 252);

      .table-thead-tr {
        border-bottom: 1px solid rgba(239, 239, 245, 1);

        .table-thead-th {
          padding: 8px;
          text-align: center;
          border-bottom: 1px solid rgba(239, 239, 245, 1);

          &.table-thead-th__sort {
            cursor: pointer;
          }

          &.table-thead-th__border {
            & ~ .table-thead-th__border {
              border-left: 1px solid rgba(239, 239, 245, 1);
            }
          }

          .table-thead-th-title {
            display: flex;
            align-items: center;
            justify-content: space-between;

            .table-thead-th-title__text {
              position: relative;
              flex: 1;
            }

            .table-thead-th-title__sort {
              color: #bfbfbf;
              display: flex;
              flex-direction: column;
              font-size: 12px;

              &__asc.active {
                color: green;
              }

              ;

              &__desc.active {
                color: red;
              }
            }
          }
        }
      }
    }

    .table-body {
      .table-body-tr {
        .table-body-td {
          padding: 8px;
          text-align: center;
          border-bottom: 1px solid rgba(239, 239, 245, 1);

          &.table-body-td-border {
            & ~ .table-body-td-border {
              border-left: 1px solid rgba(239, 239, 245, 1);;
            }
          }
        }
      }
    }
  }
}

.table-pagenation {
  margin: 12px 0;
  display: flex;
  justify-content: flex-end;

  .table-pagenation-content {
    display: flex;
    margin-right: 8px;

    .table-pagenation-content__total {
      line-height: 28px;
      font-size: 12px;
    }

    .table-pagenation-content__btn {
      border: 1px solid rgb(224, 224, 230);
      position: relative;
      cursor: pointer;
      display: flex;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      min-width: 28px;
      height: 28px;
      padding: 0 4px;
      color: rgb(51, 54, 57);
      border-radius: 4px;
      margin-left: 8px;

      // & ~ .table-pagenation-content__btn {
        
      // }

      &:hover,
      &.active {
        border-color: #18a058;
      }
    }
    .table-pagenation-content__left-btn.isFirst,
    .table-pagenation-content__right-btn.isLast {
      background-color: rgb(250, 250, 252);
        border-color: 1px solid rgb(224, 224, 230);
        user-select: none;
        cursor: not-allowed;
        &:hover {
          border-color: rgb(224, 224, 230);
        }
    }
  }
}
