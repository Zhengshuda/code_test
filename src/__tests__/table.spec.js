import { mount } from '@vue/test-utils'
import { TestTable } from '../table'

describe('Table', () => {
  const TableMount = options => mount(TestTable, options)

  test('render', () => {
    const wrapper = TableMount()
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.vm.$destroy()
    }).not.toThrow()
  })

  test('props test', () => {
    const wrapper = TableMount({
      propsData: {
        align: 'left',
      },
    })

    expect(wrapper.find('.table-body__align-left').exists()).toBeTruthy()
  })
})
