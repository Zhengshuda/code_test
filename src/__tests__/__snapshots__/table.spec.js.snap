// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render 1`] = `
"<div>
  <table>
    <thead class=\\"table-header table-header__align-left\\">
      <tr>
        <th><input type=\\"checkbox\\"></th>
        <th>111</th>
        <th>222</th>
        <th>111</th>
      </tr>
    </thead>
    <thead class=\\"table-body table-body__align-left\\">
      <tr>
        <td><input type=\\"checkbox\\"></td>
        <td>第一行</td>
        <td>第二行</td>
      </tr>
    </thead>
  </table>
</div>"
`;
